File: java/src/main/java/vsdk/toolkit/common/linealAlgebra/Vector3D.java
package vsdk.toolkit.common.linealAlgebra;

/**
Fills in x, y, and z component of a vector
*/

/**
Copies the vector v to d: d = v. They may be different vector types
*/

/**
Tolerance value for e.g. a vertex position
*/

/**
Two vectors are equal if their components are equal within the given tolerance
*/

/**
Vector difference
*/

/**
Vector sum: d = a + b
*/

/**
Scaled vector sum: d = a + s.b
*/

/**
Scalar vector product: a.b
*/

/**
Compute (T * vector) with
T = transpose[ X Y Z ] so that e.g. T.X = (1 0 0) if
X, Y, Z form a coordinate system
*/

/**
Square of vector norm: scalar product with itself
*/

/**
Norm of a vector: square root of the square norm
*/

/**
Scale a vector: d = s.v (s is a real number)
*/

/**
Scales a vector with the inverse of the real number s if not zero: d = (1/s).v
*/

/**
Normalizes a vector: scale it with the inverse of its norm
*/

/**
In product of two vectors
*/

/**
Linear combination of two vectors: d = a.v + b.w
*/

/**
Affine linear combination of two vectors: d = o + a.v + b.w
*/

/**
Triple (cross) product: d = (v3 - v2) x (v1 - v2)
*/

/**
Distance between two positions in 3D space: s = | p2 - p1 |
*/

/**
Squared distance between two positions in 3D space: s = |p2-p1|
*/

/**
Centre of two positions
*/

/**
Find the "dominant" part of the vector (eg patch-normal).
This is used to turn the point-in-polygon test into a 2D problem.
*/

public class Vector3D {
    public float x;
    public float y;
    public float z;

    public Vector3D() {
        this(0.0f, 0.0f, 0.0f);
    }

    public Vector3D(float a, float b, float c) {
        x = a;
        y = b;
        z = c;
    }

    public Vector3D(double a, double b, double c) {
        this((float)a, (float)b, (float)c);
    }

    public static Vector3D unitX() {
        return new Vector3D(1.0f, 0.0f, 0.0f);
    }

    public static Vector3D unitY() {
        return new Vector3D(0.0f, 1.0f, 0.0f);
    }

    public static Vector3D unitZ() {
        return new Vector3D(0.0f, 0.0f, 1.0f);
    }

    public Vector3D transform(Vector3D xAxis, Vector3D yAxis, Vector3D zAxis) {
        return new Vector3D(
            xAxis.dotProduct(this),
            yAxis.dotProduct(this),
            zAxis.dotProduct(this));
    }

    public float tolerance(float epsilon) {
        return epsilon * (Math.abs(x) + Math.abs(y) + Math.abs(z));
    }

    public boolean equals(Vector3D w, float epsilon) {
        return Numeric.doubleEqual(x, w.x, epsilon) &&
            Numeric.doubleEqual(y, w.y, epsilon) &&
            Numeric.doubleEqual(z, w.z, epsilon);
    }

    public CoordinateAxis dominantCoordinate() {
        float ax = Math.abs(x);
        float ay = Math.abs(y);
        float az = Math.abs(z);
        float indexValue = Math.max(ax, Math.max(ay, az));

        if (indexValue == ax) {
            return CoordinateAxis.X;
        }
        return indexValue == ay ? CoordinateAxis.Y : CoordinateAxis.Z;
    }

    public float dotProduct(Vector3D b) {
        return x * b.x + y * b.y + z * b.z;
    }

    public float norm2() {
        return x * x + y * y + z * z;
    }

    public float norm() {
        return (float)Math.sqrt(norm2());
    }

    public float distance(Vector3D p2) {
        Vector3D d = new Vector3D();
        d.subtraction(p2, this);
        return d.norm();
    }

    public float distance2(Vector3D p2) {
        Vector3D d = new Vector3D();
        d.subtraction(p2, this);
        return d.norm2();
    }

    public void set(float xParam, float yParam, float zParam) {
        x = xParam;
        y = yParam;
        z = zParam;
    }

    public void copy(Vector3D v) {
        x = v.x;
        y = v.y;
        z = v.z;
    }

    public void combine(float a, Vector3D v, float b, Vector3D w) {
        x = a * v.x + b * w.x;
        y = a * v.y + b * w.y;
        z = a * v.z + b * w.z;
    }

    public void addition(Vector3D a, Vector3D b) {
        x = a.x + b.x;
        y = a.y + b.y;
        z = a.z + b.z;
    }

    public void subtraction(Vector3D a, Vector3D b) {
        x = a.x - b.x;
        y = a.y - b.y;
        z = a.z - b.z;
    }

    public void sumScaled(Vector3D a, double s, Vector3D b) {
        x = a.x + (float)s * b.x;
        y = a.y + (float)s * b.y;
        z = a.z + (float)s * b.z;
    }

    public void scaledCopy(float s, Vector3D v) {
        x = s * v.x;
        y = s * v.y;
        z = s * v.z;
    }

    public void inverseScaledCopy(float s, Vector3D v, float epsilon) {
        float normalizedFactor = (s < -epsilon || s > epsilon) ? 1.0f / s : 1.0f;
        x = normalizedFactor * v.x;
        y = normalizedFactor * v.y;
        z = normalizedFactor * v.z;
    }

    public void normalize(float epsilon) {
        float n = norm();
        inverseScaledCopy(n, this, epsilon);
    }

    public Vector3D normalized() {
        Vector3D out = new Vector3D();
        out.copy(this);
        out.normalize(Numeric.EPSILON_FLOAT);
        return out;
    }

    public void crossProduct(Vector3D a, Vector3D b) {
        x = a.y * b.z - a.z * b.y;
        y = a.z * b.x - a.x * b.z;
        z = a.x * b.y - a.y * b.x;
    }

    public void combine3(Vector3D o, float a, Vector3D v, float b, Vector3D w) {
        x = o.x + a * v.x + b * w.x;
        y = o.y + a * v.y + b * w.y;
        z = o.z + a * v.z + b * w.z;
    }

    public void tripleCrossProduct(Vector3D v1, Vector3D v2, Vector3D v3) {
        Vector3D d1 = new Vector3D();
        Vector3D d2 = new Vector3D();
        d1.subtraction(v3, v2);
        d2.subtraction(v1, v2);
        crossProduct(d1, d2);
    }

    public void midPoint(Vector3D p1, Vector3D p2) {
        x = 0.5f * (p1.x + p2.x);
        y = 0.5f * (p1.y + p2.y);
        z = 0.5f * (p1.z + p2.z);
    }

    @Override
    public String toString() {
        return "Vector3D{" +
            "x=" + x +
            ", y=" + y +
            ", z=" + z +
            '}';
    }
}

File: typescript/src/main/vsdk/toolkit/common/linealAlgebra/Vector3D.ts
import { CoordinateAxis } from "./CoordinateAxis";
import { Numeric } from "./Numeric";

export class Vector3D {
  public x: number;
  public y: number;
  public z: number;

  public constructor(a = 0.0, b = 0.0, c = 0.0) {
    this.x = a;
    this.y = b;
    this.z = c;
  }

  public static unitX(): Vector3D {
    return new Vector3D(1.0, 0.0, 0.0);
  }

  public static unitY(): Vector3D {
    return new Vector3D(0.0, 1.0, 0.0);
  }

  public static unitZ(): Vector3D {
    return new Vector3D(0.0, 0.0, 1.0);
  }

  public transform(xAxis: Vector3D, yAxis: Vector3D, zAxis: Vector3D): Vector3D {
    return new Vector3D(
      xAxis.dotProduct(this),
      yAxis.dotProduct(this),
      zAxis.dotProduct(this)
    );
  }

  public tolerance(epsilon: number): number {
    return epsilon * (globalThis.Math.abs(this.x) + globalThis.Math.abs(this.y) + globalThis.Math.abs(this.z));
  }

  public equals(w: Vector3D, epsilon: number): boolean {
    return Numeric.doubleEqual(this.x, w.x, epsilon) &&
      Numeric.doubleEqual(this.y, w.y, epsilon) &&
      Numeric.doubleEqual(this.z, w.z, epsilon);
  }

  public dominantCoordinate(): CoordinateAxis {
    const ax = globalThis.Math.abs(this.x);
    const ay = globalThis.Math.abs(this.y);
    const az = globalThis.Math.abs(this.z);
    const indexValue = globalThis.Math.max(ax, globalThis.Math.max(ay, az));

    if (indexValue === ax) {
      return CoordinateAxis.X;
    }
    return indexValue === ay ? CoordinateAxis.Y : CoordinateAxis.Z;
  }

  public dotProduct(b: Vector3D): number {
    return this.x * b.x + this.y * b.y + this.z * b.z;
  }

  public norm2(): number {
    return this.x * this.x + this.y * this.y + this.z * this.z;
  }

  public norm(): number {
    return globalThis.Math.sqrt(this.norm2());
  }

  public distance(p2: Vector3D): number {
    const d = new Vector3D();
    d.subtraction(p2, this);
    return d.norm();
  }

  public distance2(p2: Vector3D): number {
    const d = new Vector3D();
    d.subtraction(p2, this);
    return d.norm2();
  }

  public set(xParam: number, yParam: number, zParam: number): void {
    this.x = xParam;
    this.y = yParam;
    this.z = zParam;
  }

  public copy(v: Vector3D): void {
    this.x = v.x;
    this.y = v.y;
    this.z = v.z;
  }

  public combine(a: number, v: Vector3D, b: number, w: Vector3D): void {
    this.x = a * v.x + b * w.x;
    this.y = a * v.y + b * w.y;
    this.z = a * v.z + b * w.z;
  }

  public addition(a: Vector3D, b: Vector3D): void {
    this.x = a.x + b.x;
    this.y = a.y + b.y;
    this.z = a.z + b.z;
  }

  public subtraction(a: Vector3D, b: Vector3D): void {
    this.x = a.x - b.x;
    this.y = a.y - b.y;
    this.z = a.z - b.z;
  }

  public sumScaled(a: Vector3D, s: number, b: Vector3D): void {
    this.x = a.x + s * b.x;
    this.y = a.y + s * b.y;
    this.z = a.z + s * b.z;
  }

  public scaledCopy(s: number, v: Vector3D): void {
    this.x = s * v.x;
    this.y = s * v.y;
    this.z = s * v.z;
  }

  public inverseScaledCopy(s: number, v: Vector3D, epsilon: number): void {
    const normalizedFactor = (s < -epsilon || s > epsilon) ? 1.0 / s : 1.0;
    this.x = normalizedFactor * v.x;
    this.y = normalizedFactor * v.y;
    this.z = normalizedFactor * v.z;
  }

  public normalize(epsilon: number): void {
    const n = this.norm();
    this.inverseScaledCopy(n, this, epsilon);
  }

  public normalized(): Vector3D {
    const out = new Vector3D();
    out.copy(this);
    out.normalize(Numeric.EPSILON_FLOAT);
    return out;
  }

  public crossProduct(a: Vector3D, b: Vector3D): void {
    this.x = a.y * b.z - a.z * b.y;
    this.y = a.z * b.x - a.x * b.z;
    this.z = a.x * b.y - a.y * b.x;
  }

  public combine3(o: Vector3D, a: number, v: Vector3D, b: number, w: Vector3D): void {
    this.x = o.x + a * v.x + b * w.x;
    this.y = o.y + a * v.y + b * w.y;
    this.z = o.z + a * v.z + b * w.z;
  }

  public tripleCrossProduct(v1: Vector3D, v2: Vector3D, v3: Vector3D): void {
    const d1 = new Vector3D();
    const d2 = new Vector3D();
    d1.subtraction(v3, v2);
    d2.subtraction(v1, v2);
    this.crossProduct(d1, d2);
  }

  public midPoint(p1: Vector3D, p2: Vector3D): void {
    this.x = 0.5 * (p1.x + p2.x);
    this.y = 0.5 * (p1.y + p2.y);
    this.z = 0.5 * (p1.z + p2.z);
  }

  public toString(): string {
    return `Vector3D{x=${this.x}, y=${this.y}, z=${this.z}}`;
  }
}
